### PR TITLE
Erigon 2.40.0

### PR DESCRIPTION
**Pull request by sudeepdino008**

### CODE REVIEW

File: core/types/legacy_tx.go [hunk 1]
```diff
@@ -24,22 +24,25 @@ import (
...
-	V, R, S uint256.Int     // signature values
+	Nonce   uint64             // nonce of sender account
+	Gas     uint64             // gas limit
+	To      *libcommon.Address `rlp:"nil"` // nil means contract creation
+	Value   *uint256.Int       // wei amount
+	Data    []byte             // contract invocation input data
+	V, R, S uint256.Int        // signature values
 }
 
 func (ct *CommonTx) SetFrom(from common.Address) {
```
Comment (rogarcia) [inline core/types/legacy_tx.go]: need to update this func too with *libcommon.Address right?

File: turbo/stages/headerdownload/header_algos.go [hunk 2]
```diff
@@ -525,7 +519,8 @@ func (hd *HeaderDownload) InsertHeader(hf FeedHeaderFunc, terminalTotalDifficult
 			hd.moveLinkToQueue(link, NoQueue)
 			delete(hd.links, link.hash)
 			hd.removeUpwards(link)
-			return true, false, 0, nil
+			log.Warn("[downloader] Rejected header marked as bad", "hash", link.hash, "height", link.blockHeight)
+			return true, false, 0, lastTime, nil
 		}
 		if hd.ignoreAboveHeight != nil && link.blockHeight > *hd.ignoreAboveHeight {
 			return false, true, 0, nil
```
Comment (rogarcia) [inline turbo/stages/headerdownload/header_algos.go]: ../../turbo/stages/headerdownload/header_algos.go:526:27: not enough return values
	have (bool, bool, number, nil)
	want (bool, bool, uint64, uint64, error)

File: turbo/stages/headerdownload/header_algos.go [hunk 3]
```diff
@@ -535,14 +530,14 @@ func (hd *HeaderDownload) InsertHeader(hf FeedHeaderFunc, terminalTotalDifficult
...
-					log.Debug("[Downloader] Verification failed for header", "hash", link.hash, "height", link.blockHeight, "err", err)
+					log.Debug("[downloader] Verification failed for header", "hash", link.hash, "height", link.blockHeight, "err", err)
 					hd.moveLinkToQueue(link, NoQueue)
 					delete(hd.links, link.hash)
 					hd.removeUpwards(link)
-					return true, false, 0, nil
+					return true, false, 0, lastTime, nil
 				}
 			}
 		}
```
Comment (rogarcia) [inline turbo/stages/headerdownload/header_algos.go:553]: ../../turbo/stages/headerdownload/header_algos.go:546:28: not enough return values
	have (bool, bool, number, nil)
	want (bool, bool, uint64, uint64, error)

File: core/vm/interpreter.go [hunk 4]
```diff
@@ -19,31 +19,40 @@ package vm
...
 	"bytes"
 	"hash"
-	"sync"
-	"sync/atomic"
 
-	"github.com/ledgerwatch/erigon/common"
-	"github.com/ledgerwatch/erigon/common/math"
-	"github.com/ledgerwatch/erigon/core/vm/stack"
+	"github.com/ledgerwatch/erigon-lib/chain"
+	libcommon "github.com/ledgerwatch/erigon-lib/common"
```
Comment (rogarcia) [inline core/vm/interpreter.go:25]: ../../core/vm/interpreter.go:82:19: undefined: common
../../core/vm/interpreter.go:83:19: undefined: common
../../core/vm/interpreter.go:84:19: undefined: common

**Review by rogarcia — CHANGES_REQUESTED** [review]
seems there quite a few dependency problems and doesn't compile

**Comment by sudeepdino008** [comment]
@tsutsu the evm tool tests work. Not sure how to test eth_multicall. Do you have pointers?

**Comment by noslav** [comment]
> 
@sudeepdino008 
I've some tooling/sources to make calls to multicall but on the client side, so that node PR has to be deployed to be tested against.

**Pull request closed without merging**